### PR TITLE
Implement Sigprocmask 2 and 3

### PR DESCRIPTION
**Pull request by VocalFan**
This fixes https://github.com/obhq/obliteration/issues/263 by implementing SIG_UNBLOCK (2) and SIG_SETMASK (3)

We now get an error on Syscall 608, which I'll create an issue for.

### CODE REVIEW

**Comment by VocalFan** [comment]
> Have you checked on the PS4 kernel to see if it does the same thing? Also please don't include the commits from the other PR.

Sorry, I kinda did that so I could test Sigprocmask... Seeing as well, Sigprocmask 3 only appears after 592 is implemented. And, with what I checked, Sigprocmask between PS4 and FreeBSD seems relatively unchanged.

**Comment by ultimaweapon** [comment]
Show me the piece of function on the PS4 that is corresponding to your code.

**Comment by VocalFan** [comment]
> Show me the piece of function on the PS4 that is corresponding to your code.

Time to dig through Ghidra

**Comment by VocalFan** [comment]
> Show me the piece of function on the PS4 that is corresponding to your code.

...Well, I'm an idiot. Did a search for: Guess what? String! Ahahaha, dear god do I suck at Ghidra .w.

...In general searching, sigprocmask does exist, but I cannot find kern_sigprocmask

*Is this another one of those data types you've never told me of?*

**Comment by ultimaweapon** [comment]
Navigate to the sigprocmask entry point through sysent table that I was gave you before. Then follow it using the FreeBSD as a reference.

**Comment by VocalFan** [comment]
> Navigate to the sigprocmask entry point through sysent table that I was gave you before. Then follow it using the FreeBSD as a reference.

Yep, going through Sysent 340 and hitting my head against a wall at the nested funcs... Seems good. Unblock doesn't need to set.remove sigkill or sigstop, as it would basically do nothing to the calls.

```C
void FUN_ffffffffa040d7d0(undefined8 param_1,undefined4 *param_2)

{
  long lVar1;
  long lVar2;
  undefined8 uVar3;
  undefined *puVar4;
  int iVar5;
  undefined *puVar6;
  undefined8 *in_GS_OFFSET;
  undefined local_58 [16];
  undefined local_48 [16];
  long local_38;
  
  puVar6 = local_58;
  puVar4 = local_48;
  local_38 = DAT_ffffffffa2b34010;
  lVar1 = *(long *)(param_2 + 4);
  lVar2 = *(long *)(param_2 + 2);
  if (lVar1 == 0) {
    puVar6 = (undefined *)0x0;
  }
  if (lVar2 == 0) {
    puVar4 = (undefined *)0x0;
  }
  else {
    FUN_ffffffffa0542330(*in_GS_OFFSET,"copyin",0);
    iVar5 = FUN_ffffffffa06596a0(lVar2,puVar4,0x10);
    if (iVar5 != 0) goto LAB_ffffffffa040d88c;
  }
  iVar5 = FUN_ffffffffa040d300(param_1,*param_2,puVar4,puVar6,0);
  if ((lVar1 != 0) && (iVar5 == 0)) {
    uVar3 = *(undefined8 *)(param_2 + 4);
    FUN_ffffffffa0542330(*in_GS_OFFSET,"copyout",0);
    FUN_ffffffffa06595b0(puVar6,uVar3,0x10);
  }
LAB_ffffffffa040d88c:
  if (DAT_ffffffffa2b34010 != local_38) {
                    /* WARNING: Subroutine does not return */
    FUN_ffffffffa0aafd90();
  }
  return;
}
```

**Comment by ultimaweapon** [comment]
You need to understand this function in the PS4 in order to implement #263.

**Comment by VocalFan** [comment]
> You need to understand this function in the PS4 in order to implement #263.

*Apparently everything gets lost in translation between us.*

**Comment by VocalFan** [comment]
But... I know that, Ultima.

**Comment by VocalFan** [comment]
Yeah... Still lines up, Ultima.

**Comment by ultimaweapon** [comment]
Which lines that correspond to your codes?

**Comment by VocalFan** [comment]
> Which lines that correspond to your codes?

From the code above? Function FUN_ffffffffa040d300

```

undefined8
FUN_ffffffffa040d300(long param_1,int param_2,undefined8 *param_3,undefined8 *param_4,uint param_5)

{
  undefined8 *puVar1;
  uint uVar2;
  int iVar3;
  long lVar4;
  undefined8 uVar5;
  undefined8 uVar6;
  long lVar7;
  undefined8 uVar8;
  undefined8 local_58;
  undefined8 local_50;
  uint local_48;
  uint uStack_44;
  uint local_40;
  uint uStack_3c;
  
  lVar7 = DAT_ffffffffa2b34010;
  lVar4 = *(long *)(param_1 + 8);
  if ((param_5 & 2) == 0) {
    FUN_ffffffffa06d6eb0
              (lVar4 + 0xf8,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x3d2);
  }
  if (param_4 != (undefined8 *)0x0) {
    uVar8 = *(undefined8 *)(param_1 + 0x244);
    param_4[1] = *(undefined8 *)(param_1 + 0x24c);
    *param_4 = uVar8;
  }
  uVar8 = 0;
  if (param_3 != (undefined8 *)0x0) {
    if (param_2 == 1) {
      uVar2 = *(uint *)param_3;
      *(uint *)param_3 = uVar2 & 0xfffefeff;
      uVar8 = *(undefined8 *)(param_1 + 0x244);
      uVar5 = *(undefined8 *)(param_1 + 0x24c);
      *(uint *)(param_1 + 0x244) = *(uint *)(param_1 + 0x244) | uVar2 & 0xfffefeff;
      *(uint *)(param_1 + 0x248) = *(uint *)(param_1 + 0x248) | *(uint *)((long)param_3 + 4);
      *(uint *)(param_1 + 0x24c) = *(uint *)(param_1 + 0x24c) | *(uint *)(param_3 + 1);
      *(uint *)(param_1 + 0x250) = *(uint *)(param_1 + 0x250) | *(uint *)((long)param_3 + 0xc);
      local_48 = (uint)uVar8;
      uStack_44 = (uint)((ulong)uVar8 >> 0x20);
      local_58._4_4_ = (uint)((ulong)*(undefined8 *)(param_1 + 0x244) >> 0x20);
      local_40 = (uint)uVar5;
      local_58 = CONCAT44(local_58._4_4_ & ~uStack_44,
                          ~local_48 & (uint)*(undefined8 *)(param_1 + 0x244));
      uStack_3c = (uint)((ulong)uVar5 >> 0x20);
      local_50._4_4_ = (uint)((ulong)*(undefined8 *)(param_1 + 0x24c) >> 0x20);
      local_50 = CONCAT44(local_50._4_4_ & ~uStack_3c,
                          ~local_40 & (uint)*(undefined8 *)(param_1 + 0x24c));
      iVar3 = *(int *)(lVar4 + 0x9e0);
    }
    else {
      if (param_2 != 3) {
        uVar8 = 0x16;
        if (param_2 == 2) {
          *(uint *)(param_1 + 0x244) = *(uint *)(param_1 + 0x244) & ~*(uint *)param_3;
          *(uint *)(param_1 + 0x248) = *(uint *)(param_1 + 0x248) & ~*(uint *)((long)param_3 + 4);
          *(uint *)(param_1 + 0x24c) = *(uint *)(param_1 + 0x24c) & ~*(uint *)(param_3 + 1);
          *(uint *)(param_1 + 0x250) = *(uint *)(param_1 + 0x250) & ~*(uint *)((long)param_3 + 0xc);
          FUN_ffffffffa040ca20(param_1);
          uVar8 = 0;
        }
        goto LAB_ffffffffa040d53f;
      }
      uVar2 = *(uint *)param_3;
      puVar1 = (undefined8 *)(param_1 + 0x244);
      *(uint *)param_3 = uVar2 & 0xfffefeff;
      uVar8 = *(undefined8 *)(param_1 + 0x244);
      uVar5 = *(undefined8 *)(param_1 + 0x24c);
      if ((param_5 & 1) == 0) {
        uVar6 = *param_3;
        *(undefined8 *)(param_1 + 0x24c) = param_3[1];
        *puVar1 = uVar6;
      }
      else {
        *(uint *)puVar1 = uVar2 & 0xfffefeff;
      }
      local_48 = (uint)uVar8;
      uStack_44 = (uint)((ulong)uVar8 >> 0x20);
      local_58._4_4_ = (uint)((ulong)*puVar1 >> 0x20);
      local_40 = (uint)uVar5;
      local_58 = CONCAT44(local_58._4_4_ & ~uStack_44,~local_48 & (uint)*puVar1);
      uStack_3c = (uint)((ulong)uVar5 >> 0x20);
      local_50._4_4_ = (uint)((ulong)*(undefined8 *)(param_1 + 0x24c) >> 0x20);
      local_50 = CONCAT44(local_50._4_4_ & ~uStack_3c,
                          ~local_40 & (uint)*(undefined8 *)(param_1 + 0x24c));
      FUN_ffffffffa040ca20(param_1);
      iVar3 = *(int *)(lVar4 + 0x9e0);
    }
    uVar8 = 0;
    if (iVar3 != 1) {
      FUN_ffffffffa040d590(lVar4,local_58,local_50,param_5);
      uVar8 = 0;
    }
  }
LAB_ffffffffa040d53f:
  if ((param_5 & 2) == 0) {
    FUN_ffffffffa06d7170
              (lVar4 + 0xf8,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x403);
  }
  if (DAT_ffffffffa2b34010 == lVar7) {
    return uVar8;
  }
                    /* WARNING: Subroutine does not return */
  FUN_ffffffffa0aafd90();
}
```

**Comment by VocalFan** [comment]
And it seems Param_2 is the sigprocmask SIG_ type. 1 is SIG_BLOCK, 2 is SIG_UNBLOCK, and 3 is SIG_SETMASK

**Comment by ultimaweapon** [comment]
Can you explain how you wrote your codes? How do you map the above code to your codes?

**Comment by VocalFan** [comment]
Feels like I'm writing an essay lol.

But...

`param_2 == 2` is `how == SIG_UNBLOCK`

`param_2 == 3` is `how == SIG_SETMASK`

We remove SIGKILL and SIGSTOP using set.remove, while the kernel uses an AND with `0xfffefeff` to remove them... Both achieve the same goal, I'm just following your implementation of SET_BLOCK for example on the SIGKILL and SIGSTOP removal.

For SIG_UNBLOCK, the kernel uses `param_1 + 0x244, param_1 + 0x248, param_1 + 0x24c, param_1 + 0x250` which the Rust equivalent of that is `*td.sigmask_mut() &= !set;` which, just like the kernel, it doesn't filter SIGKILL or SIGSTOP as that is unneeded.

For SIG_SETMASK, the kernel uses `param_5 & 1 == 0` as an if statement to replace `param_1` 0x244 and 0x24c with `param_3` and `param_3[1]` which is done using `*td.sigmask_mut() = set;`

We ignore the trial and error I had implementing them... As it doesn't show in git history lol.

**Comment by ultimaweapon** [comment]
What is the call to `FUN_ffffffffa040ca20` does? Why you ignored it? And what is the meaning of `param_5 & 1 == 0`? I also don't see this check on your code.

**Comment by VocalFan** [comment]
> `FUN_ffffffffa040ca20`

```c
ulong FUN_ffffffffa040ca20(undefined8 *param_1)

{
  uint uVar1;
  ulong uVar2;
  
  if (((((*(uint *)(param_1 + 0x12) == 0) && (*(int *)((long)param_1 + 0x94) == 0)) &&
       (*(int *)(param_1 + 0x13) == 0)) && (*(int *)((long)param_1 + 0x9c) == 0)) ||
     ((((~*(uint *)((long)param_1 + 0x244) & *(uint *)(param_1 + 0x12)) == 0 &&
       ((~*(uint *)(param_1 + 0x49) & *(uint *)((long)param_1 + 0x94)) == 0)) &&
      (((~*(uint *)((long)param_1 + 0x24c) & *(uint *)(param_1 + 0x13)) == 0 &&
       ((~*(uint *)(param_1 + 0x4a) & *(uint *)((long)param_1 + 0x9c)) == 0)))))) {
    uVar2 = param_1[1];
    if ((((*(uint *)(uVar2 + 0x120) == 0) && (*(int *)(uVar2 + 0x124) == 0)) &&
        ((*(int *)(uVar2 + 0x128) == 0 && (*(int *)(uVar2 + 300) == 0)))) ||
       (((((~*(uint *)((long)param_1 + 0x244) & *(uint *)(uVar2 + 0x120)) == 0 &&
          ((~*(uint *)(param_1 + 0x49) & *(uint *)(uVar2 + 0x124)) == 0)) &&
         ((~*(uint *)((long)param_1 + 0x24c) & *(uint *)(uVar2 + 0x128)) == 0)) &&
        (uVar1 = ~*(uint *)(param_1 + 0x4a) & *(uint *)(uVar2 + 300), uVar2 = (ulong)uVar1,
        uVar1 == 0)))) {
      return uVar2;
    }
  }
  FUN_ffffffffa06d7610(param_1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x24b);
  *(uint *)((long)param_1 + 0xd4) = *(uint *)((long)param_1 + 0xd4) | 0x20800;
  uVar2 = FUN_ffffffffa06d74a0
                    (*param_1,0,"W:\\Build\\J02541110\\sys\\freebsd\\sys\\kern\\kern_sig.c",0x24d);
  return uVar2;
}
```

Seems to check if no signals are pending, and if all the signals blocked in the process's signal mask are not in the process's pending signals... Basically, just checks if the process has any pending unblocked signals.

and for `((param_5 & 1) == 0)`, it seems to be a simple check. If it is 0, it directly modifies the mask with the new mask provided. But if it's anything else, it specifically adds only specific signals to the mask (Basically, set.remove SIGKILL and SIGSTOP)

**Comment by ultimaweapon** [comment]
> Seems to check if no signals are pending, and if all the signals blocked in the process's signal mask are not in the process's pending signals... Basically, just checks if the process has any pending unblocked signals.

How did you know? From the above code I really have no idea what it does.

> and for ((param_5 & 1) == 0), it seems to be a simple check. If it is 0, it directly modifies the mask with the new mask provided. But if it's anything else, it specifically adds only specific signals to the mask (Basically, set.remove SIGKILL and SIGSTOP)

Why is this check not on your code?

**Comment by VocalFan** [comment]
> > Seems to check if no signals are pending, and if all the signals blocked in the process's signal mask are not in the process's pending signals... Basically, just checks if the process has any pending unblocked signals.
> 
> How did you know? From the above code I really have no idea what it does.
> 
> > and for ((param_5 & 1) == 0), it seems to be a simple check. If it is 0, it directly modifies the mask with the new mask provided. But if it's anything else, it specifically adds only specific signals to the mask (Basically, set.remove SIGKILL and SIGSTOP)
> 
> Why is this check not on your code?

Because we don't have a param_5? We have 4 parameters.

```
pub fn sigmask(
        &mut self,
        how: i32,
        set: Option<SignalSet>,
        oset: Option<&mut SignalSet>,
    )
    ```

**Comment by ultimaweapon** [comment]
You need to figure out because this one is changing the behavior of `how = 3`.

**Comment by VocalFan** [comment]
> ((param_5 & 1) == 0)

Well, good sir, I have no absolute clue what value 5 is.

**Comment by VocalFan** [comment]
So... It seems to always be a value of 0. `iVar5 = FUN_ffffffffa040d300(param_1,*param_2,puVar4,puVar6,0);`

**Comment by VocalFan** [comment]
> So... It seems to always be a value of 0. `iVar5 = FUN_ffffffffa040d300(param_1,*param_2,puVar4,puVar6,0);`

Sigsetsize, probably.

```C
/* Prototype for the underlying system call */
       int syscall(SYS_rt_sigprocmask, int how,
                                  const kernel_sigset_t   _Nullable set,
                                  kernel_sigset_t_Nullable   oldset,
                                  size_t sigsetsize);
```
As they have it up here ^

But we don't.

```
        &self,
        how: i32,
        set: *const SignalSet,
        oset: *mut SignalSet,
 ```

**Comment by ultimaweapon** [comment]
> Sigsetsize, probably.

We already have this parameter so it is not.

**Comment by VocalFan** [comment]
> > Sigsetsize, probably.
> 
> We already have this parameter so it is not.

Or... It is. And we just have to use your assumably preexisting parameter... I mean, we have how, set, and oset, but I don't see size.

**Comment by ultimaweapon** [comment]
Sorry I forget that the PS4 does not have this parameter.

> And we just have to use your assumably preexisting parameter...

Please don't say something like this if you don't know. You can see how many parameters for each syscall in the first field of sysent structure.

**Comment by VocalFan** [comment]
> Sorry I forget that the PS4 does not have this parameter.
> 
> > And we just have to use your assumably preexisting parameter...
> 
> Please don't say something like this if you don't know. You can see how many parameters for each syscall in the first field of sysent structure.

I mean... `iVar5 = FUN_ffffffffa040d300(param_1,*param_2,puVar4,puVar6,0);` This shows that it uses 5 parameters when calling the Sigprocmask function... Though param_5 is 0.

**Comment by ultimaweapon** [comment]
`FUN_ffffffffa040d300` is not `sigprocmask`.

**Comment by VocalFan** [comment]
So, correction.

```c
kern_sigprocmask(struct thread *td, int how, sigset_t *set, sigset_t *oset,
    int flags)
```

Param_5 is flags.

**Comment by VocalFan** [comment]
https://github.com/freebsd/freebsd-src/blob/b0973980cd24ba188c83fbba2410ddb8ed6546e2/sys/kern/kern_sig.c#L968-L969

**Comment by ultimaweapon** [comment]
You can use the FreeBSD source as a reference but you must not assume the PS4 version does exactly the same thing. So what is the different? If your answer is no different then show me how you map the PS4 version to the FreeBSD version in details.

**Comment by VocalFan** [comment]
> You can use the FreeBSD source as a reference but you must not assume the PS4 version does exactly the same thing. So what is the different? If your answer is no different then show me how you map the PS4 version to the FreeBSD version in details.

Well, I have someone who'll explain it in a moment... Gonna implement it :P

**Comment by janisslsm** [comment]
> You can use the FreeBSD source as a reference but you must not assume the PS4 version does exactly the same thing. So what is the different? If your answer is no different then show me how you map the PS4 version to the FreeBSD version in details.
this is a small snippet of the function which seems to match up to the freebsd source used as reference
image

additionally to make sure that it really is kern_procsigmask, I compared it against sys_swapcontext
image
which seems to match up aswell.

https://github.com/freebsd/freebsd-src/blob/e64fe029e9d3ce476e77a478318e0c3cd201ff08/sys/kern/kern_context.c#L117-L127

**Comment by ultimaweapon** [comment]
So the implementation of `SIG_UNBLOCK` and `SIG_SETMASK` is the same?

**Comment by janisslsm** [comment]
> So the implementation of `SIG_UNBLOCK` and `SIG_SETMASK` is the same?

As far as I could tell, yes.

**Comment by ultimaweapon** [comment]
@janisslsm okay looking good. Thanks for helping @VocalFan.

File: src/kernel/src/process/mod.rs [hunk 1]
```diff
@@ -67,6 +67,18 @@ impl VProc {
...
+            }
+            SIG_SETMASK => {
+                // Remove uncatchable signals.
+                set.remove(SIGKILL);
+                set.remove(SIGSTOP);
+
+                // Replace mask.
+                *td.sigmask_mut() = set;
+            }
             v => todo!("sigprocmask with how = {v}"),
```
Comment (ultimaweapon) [inline src/kernel/src/process/mod.rs:82]: We don't need this anymore so please replace it with the correct `errno`.

**Comment by VocalFan** [comment]
Gonna implement the flag, gimme a second to test it with a build.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]
Both `SIG_UNBLOCK` and `SIG_SETMASK` invokes `signotify` at the end so please put a TODO comment about this.

**Comment by VocalFan** [comment]
> @VocalFan don't do it. We can ignore that one for now because it is always zero when invoked from the `sys_sigprocmask`.

I already did but uh...

image

Can't test it cuz your new logger might be broken.

**Comment by ultimaweapon** [comment]
> I already did but uh...

Then revert it. For the logger could you create the issue?

**Comment by VocalFan** [comment]
But so be it! Reverted.

**Comment by VocalFan** [comment]
This good?

**Comment by VocalFan** [comment]
Wait, I forgot my rustFMT-

**Review by ultimaweapon — APPROVED** [review]
The error should be `InvalidHow` but the `errno` is correct. I'll fix this later.

**Pull request merged**